Only verify the first GPG signature

phmccarty opened the pull request:
The only package in Clear Linux OS that (occasionally) has its source archive signed by multiple keys is `gnupg`, so to unblock updates for it, only verify the first signature from the .sig file.

When only one signature is present in the sig file (the common case), pass the entire sig file to `gpg --verify ...`, keeping the existing behavior.

For the implementation, I refactored `parse_key` into a more generic `parse_gpg_packets` function, and also updated tests.

Note that supporting verification of multiple signatures is much more involved, so I opened an enhancement request yesterday for that (#567).

phmccarty commented:
If this approach looks okay, I will look at adding an additional test for verification of the first GPG signature when multiple signatures are encountered.

phmccarty commented:
Hmm, I'm seeing [test errors](https://travis-ci.org/phmccarty/autospec/builds/638245154?utm_medium=notification&utm_source=email) for the travis CI in my fork... Interesting that those errors do not happen on Clear.

phmccarty commented:
The reason for test failure on Ubuntu is that "packet separator" lines (beginning with `# off=` are not present, and my changes rely on them being present.

bryteise commented:
Well hopefully the separator you are using isn't something they change to often.

bryteise commented on autospec/pkg_integrity.py:
```diff
@@ -714,33 +714,68 @@ def attempt_key_import(keyid, key_fullpath):
+            if m:
+                packets[-1]["user"] = m.group(1)
+                packets[-1]["email"] = m.group(2)
+        return packets
```
I think I'd like this to be a little more resilient to unexpected data with hopefully some only minor changes.

Could we have a packet map defined outside the loop and then only append once the offset, length, keyid, user and email are found? Then if we find what looks to be a new packet before the old one is complete we just don't add that to the packets list. Thoughts?

phmccarty replied:
Not all packets will have all of those values set. But, yes, it would be more robust to initialize the packet list first, and then populate the packet info after we determine what data is present. I will make these changes.

bryteise commented on autospec/pkg_integrity.py:
```diff
@@ -114,7 +114,22 @@ def __init__(self, pubkey=None, home=None):
+            output, err, code = self.exec_cmd(args)
+            if code != 0:
+                return None
+            num_bytes = packets[0]["length"]
```
Is it possible for length to not be defined here given how packets are currently added?

phmccarty replied:
Yes, good catch. It needs to check if the `length` key is present.

bryteise commented on autospec/pkg_integrity.py:
```diff
@@ -114,7 +114,22 @@ def __init__(self, pubkey=None, home=None):
+            with tempfile.NamedTemporaryFile(prefix="newsig-", dir=self._home, delete=False) as new_sig_file:
+                new_sig_file.write(first_sig)
+                sig_name = new_sig_file.name
+        args = self.args + ['--verify', sig_name, tarfile]
```
Do we want to do anything with the new_sig_file since it is generated like check it into the repo?

phmccarty replied:
I'm not sure, hmm.

phmccarty replied:
Given how rarely we see sig files containing multiple signatures, I would be inclined to not add the "new" sig file to the repo for now. At least its contents can always be derived from the signature file that *is* checked in.

bryteise replied:
Sounds good, we should probably cleanup the temp file then at some point.

phmccarty replied:
The temp file does get cleaned up (indirectly), because it is created in `self._home`, which is tmp directory that is cleaned up later.

phmccarty commented:
This series is ready for review again. I added two new tests for the multi-signature scenarios and fixed some error propagation issues in the `GPGCli.verify` method, thereby fixing the test for the error case when packet separator lines are not present in `gpg --list-packets`.
